balancer/randomsubsetting/randomsubsetting.go

```diff
@@ -0,0 +1,191 @@
+	return lbCfg, nil
+}
+
+func (bb) Name() string {
+	return Name
+}
+
+type subsettingBalancer struct {
+	*gracefulswitch.Balancer
+
+	logger     *internalgrpclog.PrefixLogger
+	cfg        *lbConfig
+	hashSeed   uint64
+	hashDigest *xxhash.Digest
+}
+
+func (b *subsettingBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	lbCfg, ok := s.BalancerConfig.(*lbConfig)
+	if !ok {
+		b.logger.Warningf("Received config with unexpected type %T: %v", s.BalancerConfig, s.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+
+	// Build config for the gracefulswitch balancer. It is safe to ignore
+	// JSON marshaling errors here, since the config was already validated
+	// as part of ParseConfig().
+	cfg := []map[string]any{{lbCfg.ChildPolicy.Name: lbCfg.ChildPolicy.Config}}
+	cfgJSON, _ := json.Marshal(cfg)
+	parsedCfg, err := gracefulswitch.ParseConfig(cfgJSON)
+	if err != nil {
+		return fmt.Errorf("randomsubsetting: error switching to child of type %q: %v", lbCfg.ChildPolicy.Name, err)
+	}
+	b.cfg = lbCfg
+	endpoints := resolver.State{
+		Endpoints:     b.calculateSubset(s.ResolverState.Endpoints),
+		ServiceConfig: s.ResolverState.ServiceConfig,
+		Attributes:    s.ResolverState.Attributes,
+	}
+
+	return b.Balancer.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState:  endpoints,
+		BalancerConfig: parsedCfg,
+	})
+}
+
+// calculateSubset implements the subsetting algorithm, as described in A68:
+// https://github.com/grpc/proposal/blob/master/A68-random-subsetting.md#subsetting-algorithm
+func (b *subsettingBalancer) calculateSubset(endpoints []resolver.Endpoint) []resolver.Endpoint {
+	// A helper struct to hold an endpoint and its hash.
+	type endpointWithHash struct {
+		hash uint64
+		ep   resolver.Endpoint
+	}
+
+	subsetSize := b.cfg.SubsetSize
+	if len(endpoints) <= int(subsetSize) {
+		return endpoints
+	}
+
+	hashedEndpoints := make([]endpointWithHash, len(endpoints))
+	for i, endpoint := range endpoints {
+		// For every endpoint in the list, compute a hash with previously
+		// generated seed - A68.
+		//
+		// The xxhash package's Sum64() function does not allow setting a seed.
+		// This means that we need to reset the digest with the seed for every
+		// endpoint. Without this, an endpoint will not retain the same hash
+		// across resolver updates.
+		//
+		// Note that we only hash the first address of the endpoint, as per A68.
+		b.hashDigest.ResetWithSeed(b.hashSeed)
+		b.hashDigest.Write([]byte(endpoint.Addresses[0].String()))
+		hashedEndpoints[i] = endpointWithHash{
+			hash: b.hashDigest.Sum64(),
+			ep:   endpoint,
+		}
+	}
+
+	slices.SortFunc(hashedEndpoints, func(a, b endpointWithHash) int {
+		// Note: This uses the standard library cmp package, not the
+		// github.com/google/go-cmp/cmp package. The latter is intended for
+		// testing purposes only.
+		return cmp.Compare(a.hash, b.hash)
+	})
+
+	if b.logger.V(2) {
+		b.logger.Infof("Resulting subset: %v", hashedEndpoints[:subsetSize])
+	}
```

Comment:
Can we avoid logging endpoints here and let the leaf `pickfirst` handle it? Endpoints contain attributes and a list of addresses. The addresses themselves contain more attributes and this causes log spam in case of large address lists.